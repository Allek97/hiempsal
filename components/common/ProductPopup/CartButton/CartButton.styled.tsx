import { Button } from "@components/ui";
import { css } from "@emotion/react";
import styled from "@emotion/styled";
import tw from "twin.macro";

const mainTextSize = css`
    ${tw`font-size[16.5px] letter-spacing[-0.06em]
    lg:font-size[13.5px]
    2lg:font-size[1.2vw]`}

    ${tw`2xl:font-size[18px]`}
`;

export const CartBtnWrapper = styled.div`
    ${tw`sticky bottom-10 z-index[150] overflow-hidden`}
    ${tw`height[51px] 
    lg:height[3.6vw]
    2xl:height[54px]`}

    & > div {
        ${tw`h-full`}
    }
`;

export const CartBtn = styled(Button)`
    ${tw`relative h-full mx-auto overflow-hidden bg-transparent`}
    ${mainTextSize}
    border-radius: 4px;

    & > div {
        ${tw`absolute left-0 top-0 z-20 w-full h-full background-color[#000]`}

        &:first-of-type {
            &::after {
                ${tw`content absolute right[-20px] top-0 width[30px] height[calc(100% + 20px)] background[#000]`}
                transform: rotate(-15deg);
            }
        }
        &:nth-of-type(2) {
            ${tw`flex justify-center items-center z-10`}
            @media (hover: hover) and (pointer: fine) {
                &:hover {
                    ${tw`bg-accents-9`}
                }
            }
        }
    }
`;
