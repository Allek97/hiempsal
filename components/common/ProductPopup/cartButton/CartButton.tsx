import { ButtonHTMLAttributes, FC } from "react";
import { motion, Variants } from "framer-motion";

import { AnimatedText } from "@components/utils";
import { AnimatedTextProps } from "@components/utils/animations/AnimatedText";

import { CartBtn, CartBtnWrapper } from "./CartButton.styled";

export interface Props extends ButtonHTMLAttributes<HTMLButtonElement> {
    isLoading: boolean;
    preText: string;
    loadingText: string;
}

// Variants
const variants1: Variants = {
    loading: { x: "calc(-75% - 30px)" },
    notLoading: { x: "calc(-100% - 30px)" },
};
const variants2: Variants = {
    loading: {
        backgroundColor: "#e2e2e2",
        color: "var(--secondary)",
    },
    notLoading: {
        backgroundColor: "black",
        color: "var(--primary)",
    },
};
const container: Variants = {
    visible: {
        transition: {
            staggerChildren: 0.025,
        },
    },
};

// Framer props
const boxAnimation1 = {
    initial: { x: "calc(-100% - 30px)" },
    transition: {
        duration: 0.5,
        ease: [0.19, 1, 0.22, 1],
    },
    variants: variants1,
};

const boxAnimation2 = {
    transition: {
        duration: 0.5,
        ease: [0.19, 1, 0.22, 1],
    },
    variants: variants2,
};

const CartButton: FC<Props> = ({
    isLoading,
    preText,
    loadingText,
    onClick,
}) => {
    const placeholderTextInitial: AnimatedTextProps[] = [
        { type: "heading1", text: preText },
    ];
    const placeholderTextInitialAnimated: AnimatedTextProps[] = [
        { type: "heading1", text: loadingText },
    ];

    return (
        <CartBtnWrapper
            initial={{ opacity: 0 }}
            animate={{ opacity: 1 }}
            transition={{ duration: 0.15, delay: 0.35 }}
            data-testid="cart-button"
        >
            <CartBtn
                type="submit"
                role="button"
                isHoverActive={false}
                onClick={onClick}
            >
                <motion.div
                    animate={isLoading ? "loading" : "notLoading"}
                    {...boxAnimation1}
                    data-testid="motion-x"
                />
                <motion.div
                    animate={isLoading ? "loading" : "notLoading"}
                    {...boxAnimation2}
                >
                    <motion.div
                        initial="visible"
                        animate={isLoading ? "hidden" : "visible"}
                        variants={container}
                        className="flex justify-center items-center absolute left-0 top-0 w-full h-full"
                    >
                        {placeholderTextInitial.map((item) => (
                            <AnimatedText
                                {...item}
                                key={`${item.text}-${item.type}`}
                            />
                        ))}
                    </motion.div>

                    <motion.div
                        initial="hidden"
                        animate={isLoading ? "visible" : "hidden"}
                        variants={container}
                        className="flex justify-center items-center absolute left-0 top-0 w-full h-full"
                    >
                        {placeholderTextInitialAnimated.map((item) => (
                            <AnimatedText
                                {...item}
                                key={`${item.text}-${item.type}`}
                            />
                        ))}
                    </motion.div>
                </motion.div>
            </CartBtn>
        </CartBtnWrapper>
    );
};

export default CartButton;
