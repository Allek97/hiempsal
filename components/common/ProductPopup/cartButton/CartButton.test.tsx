/* eslint-disable no-irregular-whitespace */

import { render, screen, waitFor } from "@testing-library/react";
import CartButton, { Props } from "./CartButton";

jest.mock("framer-motion", () => ({
    ...jest.requireActual("framer-motion"),
    useReducedMotion: () => true,
}));

const defaultProps: Props = {
    isLoading: false,
    loadingText: "Adding",
    preText: "Add to Cart",
};

const renderCartBtn = (props?: Props) => {
    return {
        ...render(<CartButton {...defaultProps} />),
        mockProps: { ...defaultProps, ...props },
    };
};

test("renders correctly", () => {
    renderCartBtn();

    expect(screen.getByRole("button")).toBeInTheDocument();

    const addCartHeading = screen.getByRole("heading", {
        name: /A d d t o C a r t/i,
    });
    const addingHeading = screen.getByRole("heading", {
        name: /A d d i n g/i,
    });
    expect(addCartHeading).toBeInTheDocument();
    expect(addingHeading).toBeInTheDocument();
});

test("should display add to cart before loading", async () => {
    // const { container } = renderCartBtn({ ...defaultProps, isLoading: true });
    // expect(container).toMatchSnapshot();
});
test("should display 'adding' after loading", async () => {
    // const { container } = renderCartBtn({ ...defaultProps, isLoading: true });

    // const motionBox = screen.getByTestId("motion-x");
    const component = screen.getByTestId("cart-button");

    await waitFor(() => {
        expect(component).toHaveStyle("opacity: 1;");
    });

    // expect(container).toMatchSnapshot();
});
