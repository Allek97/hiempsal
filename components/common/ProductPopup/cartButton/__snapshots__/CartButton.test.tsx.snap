// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display 'adding' after loading 1`] = `
.emotion-0 {
  position: -webkit-sticky;
  position: sticky;
  bottom: 2.5rem;
  z-index: 150;
  overflow: hidden;
  height: 51px;
}

@media (min-width: 64em) {
  .emotion-0 {
    height: 3.6vw;
  }
}

@media (min-width: 93.75em) {
  .emotion-0 {
    height: 54px;
  }
}

.emotion-0>div {
  height: 100%;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  width: 100%;
  --tw-shadow: 0 10px 15px -3px rgba(0, 0, 0, 0.1),0 4px 6px -2px rgba(0, 0, 0, 0.05);
  box-shadow: var(--tw-ring-offset-shadow, 0 0 #0000),var(--tw-ring-shadow, 0 0 #0000),var(--tw-shadow);
  border-radius: 0.125rem;
  background-color: var(--secondary);
  line-height: 1em;
  color: var(--accents-1);
  text-align: center;
  cursor: pointer;
  box-shadow: rgb(0 0 0 / 28%) 0.785217px 0.785217px 3.14087px;
  -webkit-transition: background 0.5s cubic-bezier(0.19, 1, 0.22, 1),color 0.5s cubic-bezier(0.19, 1, 0.22, 1);
  transition: background 0.5s cubic-bezier(0.19, 1, 0.22, 1),color 0.5s cubic-bezier(0.19, 1, 0.22, 1);
  position: relative;
  height: 100%;
  margin-left: auto;
  margin-right: auto;
  overflow: hidden;
  background-color: rgba(0, 0, 0, 0);
  font-size: 16.5px;
  letter-spacing: -0.06em;
  border-radius: 4px;
}

@media (min-width: 64em) {
  .emotion-3 {
    font-size: 13.5px;
  }
}

@media (min-width: 70.3125em) {
  .emotion-3 {
    font-size: 1.2vw;
  }
}

@media (min-width: 93.75em) {
  .emotion-3 {
    font-size: 18px;
  }
}

.emotion-3>div {
  position: absolute;
  left: 0px;
  top: 0px;
  z-index: 20;
  width: 100%;
  height: 100%;
  background-color: #000;
}

.emotion-3>div:first-of-type::after {
  content: "";
  position: absolute;
  right: -20px;
  top: 0px;
  width: 30px;
  height: calc(100% + 20px);
  background: #000;
  -webkit-transform: rotate(-15deg);
  -moz-transform: rotate(-15deg);
  -ms-transform: rotate(-15deg);
  transform: rotate(-15deg);
}

.emotion-3>div:nth-of-type(2) {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  z-index: 10;
}

@media (hover: hover) and (pointer: fine) {
  .emotion-3>div:nth-of-type(2):hover {
    background-color: var(--accents-9);
  }
}

<div>
  <div
    class="emotion-0 emotion-1"
    data-testid="cart-button"
    style="opacity: 1;"
  >
    <button
      class="emotion-2 emotion-3 emotion-4"
      href=""
      role="button"
      type="submit"
    >
      <div
        data-testid="motion-x"
        style="transform: translateX(calc(-100% - 30px)) translateZ(0);"
      />
      <div
        style="background-color: black;"
      >
        <div
          class="flex justify-center items-center absolute left-0 top-0 w-full h-full"
        >
          <h1
            style="line-height: normal; margin-top: 3px; font-weight: 500;"
          >
            <span
              class="animated-text-wrapper"
            >
              <span
                class="inline-block overflow-hidden"
              >
                <span
                  class="inline-block"
                  style="transform: none;"
                >
                  A
                </span>
              </span>
              <span
                class="inline-block overflow-hidden"
              >
                <span
                  class="inline-block"
                  style="transform: none;"
                >
                  d
                </span>
              </span>
              <span
                class="inline-block overflow-hidden"
              >
                <span
                  class="inline-block"
                  style="transform: none;"
                >
                  d
                </span>
              </span>
              <span
                class="inline-block overflow-hidden"
              >
                <span
                  class="inline-block"
                  style="transform: none;"
                >
                   
                </span>
              </span>
            </span>
            <span
              class="animated-text-wrapper"
            >
              <span
                class="inline-block overflow-hidden"
              >
                <span
                  class="inline-block"
                  style="transform: none;"
                >
                  t
                </span>
              </span>
              <span
                class="inline-block overflow-hidden"
              >
                <span
                  class="inline-block"
                  style="transform: none;"
                >
                  o
                </span>
              </span>
              <span
                class="inline-block overflow-hidden"
              >
                <span
                  class="inline-block"
                  style="transform: none;"
                >
                   
                </span>
              </span>
            </span>
            <span
              class="animated-text-wrapper"
            >
              <span
                class="inline-block overflow-hidden"
              >
                <span
                  class="inline-block"
                  style="transform: none;"
                >
                  C
                </span>
              </span>
              <span
                class="inline-block overflow-hidden"
              >
                <span
                  class="inline-block"
                  style="transform: none;"
                >
                  a
                </span>
              </span>
              <span
                class="inline-block overflow-hidden"
              >
                <span
                  class="inline-block"
                  style="transform: none;"
                >
                  r
                </span>
              </span>
              <span
                class="inline-block overflow-hidden"
              >
                <span
                  class="inline-block"
                  style="transform: none;"
                >
                  t
                </span>
              </span>
              <span
                class="inline-block overflow-hidden"
              >
                <span
                  class="inline-block"
                  style="transform: none;"
                >
                   
                </span>
              </span>
            </span>
          </h1>
        </div>
        <div
          class="flex justify-center items-center absolute left-0 top-0 w-full h-full"
        >
          <h1
            style="line-height: normal; margin-top: 3px; font-weight: 500;"
          >
            <span
              class="animated-text-wrapper"
            >
              <span
                class="inline-block overflow-hidden"
              >
                <span
                  class="inline-block"
                  style="transform: translateY(200%) translateZ(0);"
                >
                  A
                </span>
              </span>
              <span
                class="inline-block overflow-hidden"
              >
                <span
                  class="inline-block"
                  style="transform: translateY(200%) translateZ(0);"
                >
                  d
                </span>
              </span>
              <span
                class="inline-block overflow-hidden"
              >
                <span
                  class="inline-block"
                  style="transform: translateY(200%) translateZ(0);"
                >
                  d
                </span>
              </span>
              <span
                class="inline-block overflow-hidden"
              >
                <span
                  class="inline-block"
                  style="transform: translateY(200%) translateZ(0);"
                >
                  i
                </span>
              </span>
              <span
                class="inline-block overflow-hidden"
              >
                <span
                  class="inline-block"
                  style="transform: translateY(200%) translateZ(0);"
                >
                  n
                </span>
              </span>
              <span
                class="inline-block overflow-hidden"
              >
                <span
                  class="inline-block"
                  style="transform: translateY(200%) translateZ(0);"
                >
                  g
                </span>
              </span>
              <span
                class="inline-block overflow-hidden"
              >
                <span
                  class="inline-block"
                  style="transform: translateY(200%) translateZ(0);"
                >
                   
                </span>
              </span>
            </span>
          </h1>
        </div>
      </div>
    </button>
  </div>
</div>
`;

exports[`should display add to cart before loading 1`] = `
.emotion-0 {
  position: -webkit-sticky;
  position: sticky;
  bottom: 2.5rem;
  z-index: 150;
  overflow: hidden;
  height: 51px;
}

@media (min-width: 64em) {
  .emotion-0 {
    height: 3.6vw;
  }
}

@media (min-width: 93.75em) {
  .emotion-0 {
    height: 54px;
  }
}

.emotion-0>div {
  height: 100%;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  width: 100%;
  --tw-shadow: 0 10px 15px -3px rgba(0, 0, 0, 0.1),0 4px 6px -2px rgba(0, 0, 0, 0.05);
  box-shadow: var(--tw-ring-offset-shadow, 0 0 #0000),var(--tw-ring-shadow, 0 0 #0000),var(--tw-shadow);
  border-radius: 0.125rem;
  background-color: var(--secondary);
  line-height: 1em;
  color: var(--accents-1);
  text-align: center;
  cursor: pointer;
  box-shadow: rgb(0 0 0 / 28%) 0.785217px 0.785217px 3.14087px;
  -webkit-transition: background 0.5s cubic-bezier(0.19, 1, 0.22, 1),color 0.5s cubic-bezier(0.19, 1, 0.22, 1);
  transition: background 0.5s cubic-bezier(0.19, 1, 0.22, 1),color 0.5s cubic-bezier(0.19, 1, 0.22, 1);
  position: relative;
  height: 100%;
  margin-left: auto;
  margin-right: auto;
  overflow: hidden;
  background-color: rgba(0, 0, 0, 0);
  font-size: 16.5px;
  letter-spacing: -0.06em;
  border-radius: 4px;
}

@media (min-width: 64em) {
  .emotion-3 {
    font-size: 13.5px;
  }
}

@media (min-width: 70.3125em) {
  .emotion-3 {
    font-size: 1.2vw;
  }
}

@media (min-width: 93.75em) {
  .emotion-3 {
    font-size: 18px;
  }
}

.emotion-3>div {
  position: absolute;
  left: 0px;
  top: 0px;
  z-index: 20;
  width: 100%;
  height: 100%;
  background-color: #000;
}

.emotion-3>div:first-of-type::after {
  content: "";
  position: absolute;
  right: -20px;
  top: 0px;
  width: 30px;
  height: calc(100% + 20px);
  background: #000;
  -webkit-transform: rotate(-15deg);
  -moz-transform: rotate(-15deg);
  -ms-transform: rotate(-15deg);
  transform: rotate(-15deg);
}

.emotion-3>div:nth-of-type(2) {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  z-index: 10;
}

@media (hover: hover) and (pointer: fine) {
  .emotion-3>div:nth-of-type(2):hover {
    background-color: var(--accents-9);
  }
}

<div>
  <div
    class="emotion-0 emotion-1"
    data-testid="cart-button"
    style="opacity: 0;"
  >
    <button
      class="emotion-2 emotion-3 emotion-4"
      href=""
      role="button"
      type="submit"
    >
      <div
        data-testid="motion-x"
        style="transform: translateX(calc(-100% - 30px)) translateZ(0);"
      />
      <div>
        <div
          class="flex justify-center items-center absolute left-0 top-0 w-full h-full"
        >
          <h1
            style="line-height: normal; margin-top: 3px; font-weight: 500;"
          >
            <span
              class="animated-text-wrapper"
            >
              <span
                class="inline-block overflow-hidden"
              >
                <span
                  class="inline-block"
                  style="transform: none;"
                >
                  A
                </span>
              </span>
              <span
                class="inline-block overflow-hidden"
              >
                <span
                  class="inline-block"
                  style="transform: none;"
                >
                  d
                </span>
              </span>
              <span
                class="inline-block overflow-hidden"
              >
                <span
                  class="inline-block"
                  style="transform: none;"
                >
                  d
                </span>
              </span>
              <span
                class="inline-block overflow-hidden"
              >
                <span
                  class="inline-block"
                  style="transform: none;"
                >
                   
                </span>
              </span>
            </span>
            <span
              class="animated-text-wrapper"
            >
              <span
                class="inline-block overflow-hidden"
              >
                <span
                  class="inline-block"
                  style="transform: none;"
                >
                  t
                </span>
              </span>
              <span
                class="inline-block overflow-hidden"
              >
                <span
                  class="inline-block"
                  style="transform: none;"
                >
                  o
                </span>
              </span>
              <span
                class="inline-block overflow-hidden"
              >
                <span
                  class="inline-block"
                  style="transform: none;"
                >
                   
                </span>
              </span>
            </span>
            <span
              class="animated-text-wrapper"
            >
              <span
                class="inline-block overflow-hidden"
              >
                <span
                  class="inline-block"
                  style="transform: none;"
                >
                  C
                </span>
              </span>
              <span
                class="inline-block overflow-hidden"
              >
                <span
                  class="inline-block"
                  style="transform: none;"
                >
                  a
                </span>
              </span>
              <span
                class="inline-block overflow-hidden"
              >
                <span
                  class="inline-block"
                  style="transform: none;"
                >
                  r
                </span>
              </span>
              <span
                class="inline-block overflow-hidden"
              >
                <span
                  class="inline-block"
                  style="transform: none;"
                >
                  t
                </span>
              </span>
              <span
                class="inline-block overflow-hidden"
              >
                <span
                  class="inline-block"
                  style="transform: none;"
                >
                   
                </span>
              </span>
            </span>
          </h1>
        </div>
        <div
          class="flex justify-center items-center absolute left-0 top-0 w-full h-full"
        >
          <h1
            style="line-height: normal; margin-top: 3px; font-weight: 500;"
          >
            <span
              class="animated-text-wrapper"
            >
              <span
                class="inline-block overflow-hidden"
              >
                <span
                  class="inline-block"
                  style="transform: translateY(200%) translateZ(0);"
                >
                  A
                </span>
              </span>
              <span
                class="inline-block overflow-hidden"
              >
                <span
                  class="inline-block"
                  style="transform: translateY(200%) translateZ(0);"
                >
                  d
                </span>
              </span>
              <span
                class="inline-block overflow-hidden"
              >
                <span
                  class="inline-block"
                  style="transform: translateY(200%) translateZ(0);"
                >
                  d
                </span>
              </span>
              <span
                class="inline-block overflow-hidden"
              >
                <span
                  class="inline-block"
                  style="transform: translateY(200%) translateZ(0);"
                >
                  i
                </span>
              </span>
              <span
                class="inline-block overflow-hidden"
              >
                <span
                  class="inline-block"
                  style="transform: translateY(200%) translateZ(0);"
                >
                  n
                </span>
              </span>
              <span
                class="inline-block overflow-hidden"
              >
                <span
                  class="inline-block"
                  style="transform: translateY(200%) translateZ(0);"
                >
                  g
                </span>
              </span>
              <span
                class="inline-block overflow-hidden"
              >
                <span
                  class="inline-block"
                  style="transform: translateY(200%) translateZ(0);"
                >
                   
                </span>
              </span>
            </span>
          </h1>
        </div>
      </div>
    </button>
  </div>
</div>
`;
